src/lib/utils/typeGuards.ts

```diff
@@ -0,0 +1,322 @@
+/**
+ * Type guard utilities for webhook status handlers
+ * Provides automatic TypeScript type narrowing based on webhook status
+ */
+
+import type { WebhookMetadata } from '../core/AsyncHandler';
+import type {
+  GetProductsResponse,
+  GetProductsAsyncWorking,
+  GetProductsAsyncInputRequired,
+  GetProductsAsyncSubmitted,
+  CreateMediaBuyResponse,
+  CreateMediaBuyAsyncWorking,
+  CreateMediaBuyAsyncInputRequired,
+  CreateMediaBuyAsyncSubmitted,
+  UpdateMediaBuyResponse,
+  UpdateMediaBuyAsyncWorking,
+  UpdateMediaBuyAsyncInputRequired,
+  UpdateMediaBuyAsyncSubmitted,
+  SyncCreativesResponse,
+  SyncCreativesAsyncWorking,
+  SyncCreativesAsyncInputRequired,
+  SyncCreativesAsyncSubmitted,
+} from '../types/core.generated';
+
+// ============================================================================
+// Generic Status Check Helpers
+// ============================================================================
+
+/**
+ * Check if webhook status is 'completed'
+ */
+export function isStatusCompleted(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'completed';
+}
+
+/**
+ * Check if webhook status is 'working'
+ */
+export function isStatusWorking(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'working';
+}
+
+/**
+ * Check if webhook status is 'input-required'
+ */
+export function isStatusInputRequired(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'input-required';
+}
+
+/**
+ * Check if webhook status is 'submitted'
+ */
+export function isStatusSubmitted(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'submitted';
+}
+
+/**
+ * Check if webhook status is 'failed'
+ */
+export function isStatusFailed(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'failed';
+}
+
+/**
+ * Check if webhook status is 'rejected'
+ */
+export function isStatusRejected(metadata: WebhookMetadata): boolean {
+  return metadata.status === 'rejected';
+}
+
+// ============================================================================
+// GetProducts Type Guards
+// ============================================================================
+
+/**
+ * Type guard to check if GetProducts response is completed.
+ * Automatically narrows response type to GetProductsResponse.
+ *
+ * @example
+ * ```typescript
+ * onGetProductsStatusChange: (response, metadata) => {
+ *   if (isGetProductsCompleted(metadata, response)) {
+ *     console.log(response.products); // ✅ TypeScript knows products exists
+ *   }
+ * }
+ * ```
+ */
+export function isGetProductsCompleted(
+  metadata: WebhookMetadata,
+  response: any
+): response is GetProductsResponse {
+  return metadata.status === 'completed';
```

Comment:
these type guards accept `response: any` which kinda defeats the purpose of type safety. they trust metadata.status without actually checking if the response has the right shape. would be safer to add runtime checks like `&& "products" in response` to make sure we dont lie to typescript.